File: clients/scala/VisualClient.scala
import java.net.Socket
import java.util.Locale

object VisualClient {
  object Color {
    val black = Color(0, 0, 0)
    val red = Color(255, 0, 0)
    val green = Color(0, 255, 0)
    val blue = Color(0, 0, 255)
    def apply(r: Int, g: Int, b: Int) = new Color(r / 255, g / 255, b / 255)
  }

  case class Color(r: Float, g: Float, b: Float)
}

class VisualClient(host: String, port: Int) {

  import VisualClient._

  Locale.setDefault(new Locale("en", "US"))
  private val s = new Socket(host, port)
  private val os = s.getOutputStream

  private def sendCommand(command: String): Unit = {
    os.write((command + System.lineSeparator()).getBytes)
    os.flush()
  }

  /**
    * start queueing commands to be displayed either before main drawing
    */
  def beginPre() = sendCommand("begin pre")

  /**
    * start queueing commands to be displayed either after main drawing
    */
  def beginPost() = sendCommand("begin post")

  /**
    * mark either "pre" queue of commands as ready to be displayed
    */
  def endPre() = sendCommand("end pre")

  /**
    * mark either "post" queue of commands as ready to be displayed
    */
  def endPost() = sendCommand("end post")

  /**
    * draw a circle at (x, y) with radius r and color color
    */
  def circle(x: Float, y: Float, r: Float, color: Color = Color.black) = sendCommand(f"circle $x%1.1f $y%1.1f $r%1.1f ${color.r}%1.1f ${color.g}%1.1f ${color.b}%1.1f")

  /**
    * draw a filled circle at (x, y) with radius r and color color
    */
  def fillCircle(x: Float, y: Float, r: Float, color: Color = Color.black) = sendCommand(f"fill_circle $x%1.1f $y%1.1f $r%1.1f ${color.r}%1.1f ${color.g}%1.1f ${color.b}%1.1f")

  /**
    * draw a rect with corners at (x, y) to (x, y) with color color
    */
  def rect(x1: Float, y1: Float, x2: Float, y2: Float, color: Color = Color.black) = sendCommand(f"rect $x1%1.1f $y1%1.1f $x2%1.1f $y2%1.1f ${color.r}%1.1f ${color.g}%1.1f ${color.b}%1.1f")

  /**
    * draw a filled rect with corners at (x1, y1) to (x2, y2) with color color
    */
  def fillRect(x1: Float, y1: Float, x2: Float, y2: Float, color: Color = Color.black) = sendCommand(f"fill_rect $x1%1.1f $y1%1.1f $x2%1.1f $y2%1.1f ${color.r}%1.1f ${color.g}%1.1f ${color.b}%1.1f")

  /**
    * draw a line from (x1, y1) to (x2, y2) with color color
    */
  def line(x1: Float, y1: Float, x2: Float, y2: Float, color: Color = Color.black) = sendCommand(f"line $x1%1.1f $y1%1.1f $x2%1.1f $y2%1.1f ${color.r}%1.1f ${color.g}%1.1f ${color.b}%1.1f")

  /**
    * show msg at coordinates (x, y) with color color
    */
  def text(x: Float, y: Float, msg: String, color: Color = Color.black) = sendCommand(f"text $x%1.1f $y%1.1f $msg ${color.r}%1.1f ${color.g}%1.1f ${color.b}%1.1f")

  def stop() = {
    os.close()
  }
}

File: clients/csharp/VisualClient.cs
using System;
using System.Globalization;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace Com.CodeGame.CodeWizards2016.DevKit.CSharpCgdk
{
    public sealed class VisualClient
    {
        private const int BUFFER_SIZE_BYTES = 1 << 20;

        private readonly TcpClient _client;
        private readonly BinaryWriter _writer;

        public VisualClient(string host, int port)
        {
            _client = new TcpClient(host, port)
            {
                SendBufferSize = BUFFER_SIZE_BYTES,
                ReceiveBufferSize = BUFFER_SIZE_BYTES,
                NoDelay = true
            };

            _writer = new BinaryWriter(_client.GetStream());

            CultureInfo newCInfo = (CultureInfo) Thread.CurrentThread.CurrentCulture.Clone();
            newCInfo.NumberFormat.NumberDecimalSeparator = ".";
            Thread.CurrentThread.CurrentCulture = newCInfo;
        }

        //---------------------------------------------------------------------
        // Public
        //---------------------------------------------------------------------

        /// <summary>
        /// Start queueing commands to be displayed either before main drawing
        /// </summary>
        public void BeginPre()
        {
            SendCommand("begin pre");
        }

        /// <summary>
        /// Start queueing commands to be displayed either after main drawing
        /// </summary>
        public void BeginPost()
        {
            SendCommand("begin post");
        }

        /// <summary>
        /// start queueing commands to be displayed on the absolute coordinates
        /// </summary>
        public void beginAbs() {
            SendCommand("begin abs");
        }

        /// <summary>
        /// Mark either "pre" queue of commands as ready to be displayed
        /// </summary>
        public void EndPre()
        {
            SendCommand("end pre");
        }

        /// <summary>
        /// Mark either "post" queue of commands as ready to be displayed
        /// </summary>
        public void EndPost()
        {
            SendCommand("end post");
        }

        /// <summary>
        // mark either "abs" queue of commands as ready to be displayed
        /// <summary>
        public void endAbs() {
           SendCommand("end abs");
       }


        /// <summary>
        /// Draw a circle at (x, y) with radius r and color color
        /// </summary>
        public void Circle(double x, double y, float radius, float r = 0f, float g = 0f, float b = 0f)
        {
            ValidateColor(r, g, b);
        	
            SendCommand($"circle {x} {y} {radius} {r} {g} {b}");
        }

        /// <summary>
        /// Draw a filled circle at (x, y) with radius r and color color
        /// </summary>
        public void FillCircle(double x, double y, float radius, float r = 0f, float g = 0f, float b = 0f)
        {
            ValidateColor(r, g, b);

            SendCommand($"fill_circle {x} {y} {radius} {r} {g} {b}");
        }

        /// <summary>
        /// Draw a rect with corners at (x, y) to (x, y) with color color
        /// </summary>
        public void Rect(double x1, double y1, double x2, double y2, float r = 0f, float g = 0f, float b = 0f)
        {
            ValidateColor(r, g, b);

            SendCommand($"rect {x1} {y1} {x2} {y2} {r} {g} {b}");
        }

        /// <summary>
        /// Draw a filled rect with corners at (x1, y1) to (x2, y2) with color color
        /// </summary>
        public void FillRect(double x1, double y1, double x2, double y2, float r = 0f, float g = 0f, float b = 0f)
        {
            ValidateColor(r, g, b);

            SendCommand($"fill_rect {x1} {y1} {x2} {y2} {r} {g} {b}");
        }

        /// <summary>
        /// Draw a line from (x1, y1) to (x2, y2) with color color
        /// </summary>
        public void Line(double x1, double y1, double x2, double y2, float r = 0f, float g = 0f, float b = 0f)
        {
            ValidateColor(r, g, b);

            SendCommand($"line {x1} {y1} {x2} {y2} {r} {g} {b}");
        }

        /// <summary>
        /// Show msg at coordinates (x, y) with color color
        /// </summary>
        public void Text(double x, double y, string msg, float r = 0f, float g = 0f, float b = 0f)
        {
            ValidateColor(r, g, b);

            SendCommand($"text {x} {y} {msg} {r} {g} {b}");
        }

        /// <summary>
        /// draw a arc with center at (x, y), radius radius and angle arcAngle, started from startAngle with color color, angles in radians
        /// < /summary>
        public void Arc(double x, double y, double radius, double startAngle, double arcAngle, float r = 0f, float g = 0f, float b = 0f) {
	    ValidateColor(r, g, b);
            SendCommand($"arc {x} {y} {radius} {startAngle} {arcAngle} {r} {g} {b}");
        }

        /// <summary>
        /// draw a filled arc with center at (x, y), radius radius and angle arcAngle, started from startAngle with color color, angles in radians
        /// </summary>
        public void FillArc(double x, double y, double radius, double startAngle, double arcAngle, float r = 0f, float g = 0f, float b = 0f) {
	    ValidateColor(r, g, b);
            SendCommand($"fill_arc {x} {y} {radius} {startAngle} {arcAngle} {r} {g} {b}");
        }


        //---------------------------------------------------------------------
        // Helpers
        //---------------------------------------------------------------------

        private void SendCommand(string command)
        {
            WriteString(command + "\n");
            _writer.Flush();
        }

        private void WriteString(string value)
        {
            if (value == null)
            {
                WriteInt(-1);
                return;
            }

            var bytes = Encoding.UTF8.GetBytes(value);
            _writer.Write(bytes, 0, bytes.Length);
        }

        private void WriteInt(int value)
        {
            _writer.Write(value);
        }

        public void Close()
        {
            _client.Close();
        }

        private static void ValidateColor(float r, float g, float b)
        {
            const string outOfRangeMessage = "The color value must be in range from 0.0 to 1.0 .";
            if (r < 0 || r > 1)
            {
                throw new ArgumentOutOfRangeException(nameof(r), outOfRangeMessage);
            }

            if (g < 0 || g > 1)
            {
                throw new ArgumentOutOfRangeException(nameof(g), outOfRangeMessage);
            }

            if (b < 0 || b > 1)
            {
                throw new ArgumentOutOfRangeException(nameof(b), outOfRangeMessage);
            }
        }
    }
}